docs/hacking.md

```diff
@@ -0,0 +1,176 @@
+# Hacking the Carp Compiler
+
+This doc contains various tips and tricks, notes, explanations and examples
+that can help you make changes to the Carp compiler. Be forewarned that it's
+not an exhuasitve guide book, and likely will remain a hodgepodge of
+accumulated remarks, observations and hints contributed by people that have
+modified the compiler in the past.
+
+> Note: General familiarity with compilers and compilation terminology is
+> assumed.
+
+## Structure
+
+The Carp compiler source lives in the `src/` directory. Carp is, roughly
+speaking, oragnized into four primary passes or components:
+
+![carp compiler phases](../compiler-passes.svg)
+
+Each source file plays a part in one or more components/phases in the compiler.
+The sections below breifly describe the purpose of each stage and list
+important source files. You can use these sections to get a rough idea of what
+files you might need to edit in order to alter the functionality of a
+particular phase.
+ 
+> Note: Some sources contain defintions that are important or used in pretty
+> much every phase of the compiler, in result some files may appear more than
+> once in the sections below.
+
+### Parsing
+
+The parsing phase translates `.carp` source files into abstract syntax trees
+(AST). In carp, AST nodes are represented using an abstract data type called
+`XObj`. `XObj`s are ubiquitous across the compiler and are used in serveral
+different phases and contexts. Every `XObj` consists of:
+
+- An `Obj` which is the representation of some carp source code as an abstract
+  data type
+- `Info`: which contains additional information about the source code that
+  generated an `Obj` (e.g. its location in a source file)
+- `Ty`: An valid carp Type for the `Obj`, as determined by the [type
+  system](#type-system).
+
+The following sources are important for parsing:
+
+- `Parsing.hs` -- parsing logic that translates carp source code into abstract
+  syntax.
+- `XObj.hs` -- defines the valid Carp AST nodes.
+
+### Dynamic Evaluator
+
+As stated in the [Macro guide](Macros.md#inner-workings) the dynamic evaluator
+is the central component in the compiler. As the name suggests, the evaluator
+evaluates parsed carp code (`XObjs`) and prepares it for
+[emission](#code-emission). Evaluation entails:
+
+- Expanding macros and dynamic functions
+- Resolving bindings to other forms
+- Requesting type inference for forms
+- Requesting borrow checking for forms
+
+In addition to the `XObjs` corresponding to the source file being compiled, the
+evauator relies on a `Context`--`Context` is a global object that contains
+state for the compiler. The compiler's `Context` is comprised of several
+environments, defined by the `Env` type--which hold references to known
+bindings. Different environments are used by different pahses of the compiler
```

Comment:
Should read `phases`.

Reply:
Thanks! Yep, didn't have spellcheck turned on for the initial commit--it was riddled with typos!